Formula/dsda-doom.rb

```diff
@@ -0,0 +1,58 @@
+class DsdaDoom < Formula
+  desc "Fork of prboom+ with a focus on speedrunning"
+  homepage "https://github.com/kraflab/dsda-doom"
+  url "https://github.com/kraflab/dsda-doom/archive/refs/tags/v0.24.3.tar.gz"
+  sha256 "d4cfc82eea029068329d6b6a2dcbe0b316b31a60af12e6dc5ad3e1d2c359d913"
+  license "GPL-2.0-only"
+
+  head do
+    url "https://github.com/kraflab/dsda-doom.git", branch: "master"
+  end
+
+  depends_on "cmake" => :build
+  depends_on "fluid-synth"
+  depends_on "libvorbis"
+  depends_on "mad"
+  depends_on "portmidi"
+  depends_on "sdl2"
+  depends_on "sdl2_image"
+  depends_on "sdl2_mixer"
+
+  on_linux do
+    depends_on "pcre"
+    depends_on "zlib"
+  end
+
+  def install
+    system "cmake", *std_cmake_args,
+                    "-S", "prboom2",
+                    "-B", "build",
+                    "-DBUILD_GL=ON",
+                    "-DWITH_DUMB=OFF",
+                    "-DWITH_IMAGE=ON",
+                    "-DWITH_FLUIDSYNTH=ON",
+                    "-DWITH_MAD=ON",
+                    "-DWITH_PCRE=ON",
+                    "-DWITH_PORTMIDI=ON",
+                    "-DWITH_VORBISFILE=ON",
+                    "-DWITH_ZLIB=ON"
+    system "cmake", "--build", "build", "--config", "Release"
+    system "cmake", "--install", "build", "--config", "Release"
+  end
+
+  def caveats
+    <<~EOS
+      Note that this formula only installs a Doom game engine, and no
+      actual levels. The original Doom levels are still under copyright,
+      you can find the shareware versions on:
+      https://www.doomworld.com/classicdoom/info/shareware.php
+      Place your WAD files in:
+      #{prefix}/share/games/doom
```

Comment:
This directory gets wiped out when `brew` upgrades `dsda-doom`. Can we point it somewhere more persistent?

Reply:
Usually all source ports would look for WADs in the `<prefix>/share/games/doom` directory.

I noticed that `crispy-doom` and `chocolate-doom` formulas keep their search directory to `/usr/local/share/games/doom` even with the M1 bottle. Would it be okay to set it to this path?

Reply:
It looks like we can configure this with `DOOMWADDIR`. Maybe something like
```ruby
"-DDOOMWADDIR=#{HOMEBREW_PREFIX}/share/games/doom"
```

Reply:
I'm having issues with the installation using a custom `DOOMWADDIR`, on my M1 mac, I can't seem to be able to create the `HOMEBREW_PREFIX/share/games/doom` directory structure:
- manually with the `(HOMEBREW_PREFIX/"share/games/doom").mkpath` in the `install` method,
- during the `cmake --install` step,
- after either failed using the interactive shell.

On the other hand, the directory creation does not fail in Linux nor if I run a `mkdir` from the terminal. Is this an intended restriction or a misconfiguration on my end (`brew doctor` reports no issue)?

Reply:
Ah, this is because of Homebrew's macOS sandbox, which limits the locations that formulae can write to when doing `brew install`.

Does CMake try to install anything into `DOOMWADDIR`, or is it just trying to create the directory?

If it's the latter, I wonder if we can get CMake to skip the creation of `DOOMWADDIR`, and preferably contribute this upstream as a build option.

I imagine this will be useful for users who want to install this software but don't manage the location where the wad files will be installed.

Reply:
Unfortunately, CMake installs the `dsda-doom.wad` to this directory which is required at runtime.

Alternatively, we could leave the default for `DOOMWADDIR` and tell the user to set the `DOOMWADPATH` environment variable:
```
you can find the shareware versions on:
https://www.doomworld.com/classicdoom/info/shareware.php
Add your WAD folder to the `DOOMWADPATH`
```

This is probably the best option as it is compatible with crispy and chocolate Doom.

Reply:
Ok, thanks for the explanation. I wonder if we should just symlink `share/"games/doom"` to `HOMEBREW_PREFIX/"share/games/doom"`. Being able to put `wad` files in there is much nicer than relying on an environment variable. (Though you could do that too if you want to put them elsewhere.)

We'd have to handle `dsda-doom.wad`, but that's not really a huge impediment.

Reply:
Currently, `games/doom` gets symlinked to `HOMEBREW_PREFIX/share` and users can add their own WADs to it. The issue is as you mentioned previously, when a reinstall/upgrade happens, the folder gets wiped out.

Looking at the source, additional search path are hardcoded:
```c
    {DOOMWADDIR}, // build-time configured DOOMWADDIR
    {NULL, "doom", "HOME"}, // ~/doom
    {NULL, NULL, "HOME"}, // ~
    {"/usr/local/share/games/doom"},
    {"/usr/share/games/doom"},
    {"/usr/local/share/doom"},
    {"/usr/share/doom"},
```

For x86_64 mac, the `/usr/local/share/games/doom` would correspond to `HOMEBREW_PREFIX/share/games/doom` (with the default homebrew installation). So we may need to warn users through caveats to *not* place their WADs there as it may be overwritten by upgrades.

Reply:
Even more reason to use my symlink suggestion:

This will allow users to place the WADs into `$HOMEBREW_PREFIX/share/games/doom` without having to worry about them ever getting wiped out, even if `HOMEBREW_PREFIX == "/usr/local"`.

Reply:
I've pushed a change that implements my suggestion. Please test it out and let me know what you think.

Reply:
Thank you for your help!

Unfortunately, this change does not seem to work. When running `brew reinstall --build-from-source dsda-doom`:
- The `post_install` method raises an exception in `mkpath` (`Errno::EEXIST: File exists @ dir_s_mkdir - /opt/homebrew/share/games/doom`).
- The content of `HOMEBREW_PREFIX/share/games/doom` is replaced by a single broken symlink:
```
% ls -l /opt/homebrew/share/games/doom
/opt/homebrew/share/games/doom ⇒ ../../../../../share/games/doom
```

I also noticed that with this patch if a WAD is added to `HOMEBREW_PREFIX/share/games/doom`, running `brew uninstall dsda-doom` add this WAD to its file count:
```
% brew uninstall dsda-doom
Uninstalling /opt/homebrew/Cellar/dsda-doom/0.24.3... (6 files, 6.5MB)
...
% cp DOOM.WAD /opt/homebrew/share/games/doom
% brew uninstall dsda-doom
Uninstalling /opt/homebrew/Cellar/dsda-doom/0.24.3... (7 files, 18MB)
```

Reply:
Oh yes, I see. The symlink inside `prefix` needs to be created in `post_install` too.

Reply:
Ok, try now?

Reply:
Works great, thank you!